File: M/318-maximum-product-of-word-lengths/main.go
package main

import "fmt"

func maxProduct(words []string) int {
	buf := make([]int, len(words))
	for i, v := range words {
		s := 0
		for _, c := range v {
			s |= 1 << int(c-'a')
		}
		buf[i] = s
	}
	max := 0
	for i := 0; i < len(buf); i++ {
		for j := i + 1; j < len(buf); j++ {
			if buf[i]&buf[j] == 0 && max < len(words[i])*len(words[j]) {
				max = len(words[i]) * len(words[j])
			}
		}
	}
	return max
}

func maxProductV2(words []string) int {
	buf := make([][]int, len(words))
	for i, v := range words {
		buf[i] = make([]int, 26)
		for _, c := range v {
			buf[i][c-'a']++
		}
	}
	max := 0
	for i := 0; i < len(buf); i++ {
		for j := i + 1; j < len(buf); j++ {
			isValid := true
			for n := 0; n < 26; n++ {
				if buf[i][n] > 0 && buf[j][n] > 0 {
					isValid = false
					break
				}
			}
			if isValid && max < len(words[i])*len(words[j]) {
				max = len(words[i]) * len(words[j])
			}
		}
	}
	return max
}

func main() {
	fmt.Println("a")
}

File: M/318-maximum-product-of-word-lengths/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn max_product(words: Vec<String>) -> i32 {
    let mut buf: HashMap<i32, i32> = HashMap::new();
    words.iter().for_each(|word| {
      let mut mask: i32 = 0;
      word.as_bytes().iter().for_each(|&c| {
        mask |= 1 << (c - b'a') as i32;
      });
      buf.insert(
        mask,
        std::cmp::max(*buf.get(&mask).unwrap_or(&0), (word.len() as i32)),
      );
    });
    let mut max: i32 = 0;
    for (k1, v1) in buf.iter() {
      for (k2, v2) in buf.iter() {
        if k1 & k2 == 0 {
          max = std::cmp::max(max, v1 * v2);
        }
      }
    }
    max
  }
}
